File: 4.search and sorting/102.first and last/code.cpp
Ques:https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/description/

Code:

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> result = {-1, -1};
        int left = binarySearch(nums, target, true);
        int right = binarySearch(nums, target, false);
        result[0] = left;
        result[1] = right;
        return result;        
    }

    int binarySearch(vector<int>& nums, int target, bool isSearchingLeft) {
        int left = 0;
        int right = nums.size() - 1;
        int idx = -1;
        
        while (left <= right) {
            int mid = left + (right - left) / 2;
            
            if (nums[mid] > target) {
                right = mid - 1;
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else {
                idx = mid;
                if (isSearchingLeft) {
                    right = mid - 1; // search left half
                } else {
                    left = mid + 1;  // search right half
                }
            }
        }
        
        return idx;
    }    
};

// Driver code
int main() {
    Solution sol;
    vector<int> nums = {5, 7, 7, 8, 8, 10};
    int target = 8;

    vector<int> result = sol.searchRange(nums, target);
    cout << "First and last positions of " << target << " are: ";
    cout << "[" << result[0] << ", " << result[1] << "]" << endl;

    return 0;
}


Sample Input
cpp
Copy
Edit
arr[] = {1, 2, 2, 2, 2, 3, 4, 7, 8, 8}
x = 2
✅ Output
Copy
Edit
1 4

Time & Space Complexity
Metric	Value
Time Complexity	O(log n)
Space Complexity	O(1)

File: 12.graph/364.flood fill/code.py
Ques:https://leetcode.com/problems/flood-fill/description/


Yes ✅ — your Flood Fill solution here is using Depth-First Search (DFS).
Both DFS and BFS work for Flood Fill.
DFS is preferred because it is simpler, uses less auxiliary data structures, and does not require shortest-path guarantees.

Code:

✅ Approach: Depth-First Search (DFS)

Start from (sr, sc)

Move in 4 directions (up, down, left, right)

Recolor only pixels with the same original color

Recolored pixels act as visited

🧾 Code (with Sample Input & Output):

✅ Approach

Depth-First Search (DFS)

Start DFS from (sr, sc)

Move in 4 directions (up, down, left, right)

Recolor only pixels matching the original color

Recolored pixels automatically act as visited

Code:


def func(r,c,image,pc,n,m,old_color):
    
    if r<0 or r>=n or c<0 or c>=m:
        return
    if image[r][c]!=old_color:
        return
    image[r][c]=pc
    func(r+1,c,image,pc,n,m,old_color)
    func(r-1,c,image,pc,n,m,old_color)
    func(r,c+1,image,pc,n,m,old_color)
    func(r,c-1,image,pc,n,m,old_color)


def solv(r,c,image,pc):
    n=len(image)
    m=len(image[0])
    if image[r][c] == pc:
        return
    old_color = image[r][c]
    
    func(r,c,image,pc,n,m,old_color)
    
    

image = [
    [1, 1, 1],
    [1, 1, 0],
    [1, 0, 1]
]

sc=1
sr=1

pc=2

solv(sr,sc,image,pc)
for i in range(len(image)):
    for j in range(len(image[i])):
        print(image[i][j])
    
    print()

▶️ Sample Output
[2, 2, 2]
[2, 2, 0]
[2, 0, 1]

⏱️ Time Complexity
O(N × M)


In the worst case, every pixel is visited once.

🧠 Space Complexity
O(N × M)


Due to recursion stack in the worst case (entire grid filled)